File: sdk/ts-sdk/src/signing.ts
/**
 * Ed25519 signing utilities for Demiurge transactions
 */

import { signAsync, getPublicKeyAsync } from '@noble/ed25519';

/**
 * Hash bytes using SHA-256 (Web Crypto API)
 */
async function sha256Hash(data: Uint8Array): Promise<Uint8Array> {
  const dataCopy = new Uint8Array(data);
  const hashBuffer = await crypto.subtle.digest('SHA-256', dataCopy);
  return new Uint8Array(hashBuffer);
}

/**
 * Convert hex string to bytes
 */
function hexToBytes(hex: string): Uint8Array {
  const cleanHex = hex.startsWith('0x') ? hex.slice(2) : hex;
  const bytes = new Uint8Array(cleanHex.length / 2);
  for (let i = 0; i < cleanHex.length; i += 2) {
    bytes[i / 2] = parseInt(cleanHex.slice(i, i + 2), 16);
  }
  return bytes;
}

/**
 * Convert bytes to hex string
 */
function bytesToHex(bytes: Uint8Array): string {
  return Array.from(bytes)
    .map((b) => b.toString(16).padStart(2, '0'))
    .join('');
}

/**
 * Derive public key (address) from private key
 */
export async function deriveAddress(privateKeyHex: string): Promise<string> {
  const privateKey = hexToBytes(privateKeyHex);
  const publicKey = await getPublicKeyAsync(privateKey);
  return '0x' + bytesToHex(publicKey);
}

/**
 * Sign a message with Ed25519
 */
export async function signMessage(
  message: Uint8Array,
  privateKeyHex: string
): Promise<string> {
  const privateKey = hexToBytes(privateKeyHex);
  const signature = await signAsync(message, privateKey);
  return bytesToHex(signature);
}

/**
 * Sign a transaction
 * 
 * The transaction should be serialized without signature (for signing),
 * then the signature is computed and added.
 */
export async function signTransaction(
  txBytes: Uint8Array,
  privateKeyHex: string
): Promise<string> {
  // Hash the transaction bytes for signing
  const messageHash = await sha256Hash(txBytes);
  return signMessage(messageHash, privateKeyHex);
}

/**
 * Sign a transaction hex string
 */
export async function signTransactionHex(
  unsignedTxHex: string,
  privateKeyHex: string
): Promise<string> {
  const txBytes = hexToBytes(unsignedTxHex);
  return signTransaction(txBytes, privateKeyHex);
}

File: sdk/rust-sdk/src/signing.rs
/*!
 * Ed25519 signing utilities for Demiurge transactions
 */

use crate::error::{DemiurgeError, Result};
use ed25519_dalek::{Signer, SigningKey, VerifyingKey};
use sha2::{Digest, Sha256};

/// Derive public key (address) from private key
pub fn derive_address(private_key: &[u8; 32]) -> Result<String> {
    let signing_key = SigningKey::from_bytes(private_key);
    let verifying_key: VerifyingKey = signing_key.verifying_key();
    Ok(format!("0x{}", hex::encode(verifying_key.as_bytes())))
}

/// Sign a message with Ed25519
pub fn sign_message(message: &[u8], private_key: &[u8; 32]) -> Result<String> {
    let signing_key = SigningKey::from_bytes(private_key);
    let signature = signing_key.sign(message);
    Ok(hex::encode(signature.to_bytes()))
}

/// Sign a transaction (hashes the transaction bytes first)
pub fn sign_transaction(tx_bytes: &[u8], private_key: &[u8; 32]) -> Result<String> {
    let mut hasher = Sha256::new();
    hasher.update(tx_bytes);
    let hash = hasher.finalize();
    sign_message(&hash, private_key)
}

/// Convert hex string to bytes
pub fn hex_to_bytes(hex: &str) -> Result<Vec<u8>> {
    let clean_hex = hex.strip_prefix("0x").unwrap_or(hex);
    hex::decode(clean_hex)
        .map_err(|e| DemiurgeError::Other(format!("Invalid hex string: {}", e)))
}

/// Convert hex string to 32-byte array (for private keys)
pub fn hex_to_bytes_32(hex: &str) -> Result<[u8; 32]> {
    let bytes = hex_to_bytes(hex)?;
    if bytes.len() != 32 {
        return Err(DemiurgeError::Other("Expected 32 bytes".to_string()));
    }
    let mut array = [0u8; 32];
    array.copy_from_slice(&bytes);
    Ok(array)
}

